File: src/remote.rs
use std::{fs, process::Command, time::Duration};

use anyhow::{Context, Result, anyhow};
use semver::Version;
use serde::{Deserialize, Serialize};

use crate::{
    cache::Cache,
    github::{GitHubRepo, github_token},
    util::{atomic_write, command_output, http_client},
};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub(crate) struct RemoteVersion {
    pub(crate) version: Version,
    pub(crate) revision: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct RemoteVersionsCache {
    location: String,
    versions: Vec<RemoteVersion>,
}

pub(crate) fn remote_versions(location: &str, cache: &Cache) -> Result<Vec<RemoteVersion>> {
    if let Some(versions) = read_cached_remote_versions(cache, location)? {
        return Ok(versions);
    }

    let _lock = cache.lock("remote-versions", location)?;
    if let Some(versions) = read_cached_remote_versions(cache, location)? {
        return Ok(versions);
    }

    let versions = fetch_remote_versions(location)?;
    write_cached_remote_versions(cache, location, &versions)?;
    Ok(versions)
}

fn fetch_remote_versions(location: &str) -> Result<Vec<RemoteVersion>> {
    // `git ls-remote --tags` returns every tag with its commit SHA in a single
    // round trip. The GitHub tags API paginates (100 tags per page) and adds
    // JSON parsing overhead, so we only fall back to it when ls-remote fails
    // (e.g. SSH-only auth without git available).
    if let Ok(versions) = git_remote_versions(location) {
        if !versions.is_empty() {
            return Ok(versions);
        }
    }
    if let Ok(github) = GitHubRepo::parse(location) {
        return github_remote_versions(&github);
    }
    Ok(Vec::new())
}

fn read_cached_remote_versions(
    cache: &Cache,
    location: &str,
) -> Result<Option<Vec<RemoteVersion>>> {
    let path = cache.remote_versions_path(location);
    if !path.exists() {
        return Ok(None);
    }

    let metadata = fs::metadata(&path)?;
    let modified = metadata.modified()?;
    if modified.elapsed().unwrap_or(Duration::MAX) > Duration::from_secs(60 * 60) {
        return Ok(None);
    }

    let file = fs::File::open(&path)?;
    let cached: RemoteVersionsCache = serde_json::from_reader(file)
        .with_context(|| format!("failed to parse {}", path.display()))?;
    if cached.location != location {
        return Ok(None);
    }
    Ok(Some(cached.versions))
}

fn write_cached_remote_versions(
    cache: &Cache,
    location: &str,
    versions: &[RemoteVersion],
) -> Result<()> {
    let path = cache.remote_versions_path(location);
    let bytes = serde_json::to_vec_pretty(&RemoteVersionsCache {
        location: location.to_string(),
        versions: versions.to_vec(),
    })?;
    atomic_write(&path, &bytes)?;
    Ok(())
}

fn github_remote_versions(repo: &GitHubRepo) -> Result<Vec<RemoteVersion>> {
    #[derive(Deserialize)]
    struct Tag {
        name: String,
        commit: Commit,
    }
    #[derive(Deserialize)]
    struct Commit {
        sha: String,
    }

    let client = http_client();
    let mut page = 1;
    let mut versions = Vec::new();
    loop {
        let url = format!(
            "https://api.github.com/repos/{}/{}/tags?per_page=100&page={page}",
            repo.owner, repo.repo
        );
        let mut request = client.get(url);
        if let Some(token) = github_token() {
            request = request.bearer_auth(token);
        }
        let tags: Vec<Tag> = request.send()?.error_for_status()?.json()?;
        if tags.is_empty() {
            break;
        }
        for tag in tags {
            if let Some(version) = parse_swift_tag_version(&tag.name) {
                versions.push(RemoteVersion {
                    version,
                    revision: tag.commit.sha,
                });
            }
        }
        page += 1;
    }
    Ok(versions)
}

fn git_remote_versions(location: &str) -> Result<Vec<RemoteVersion>> {
    let output = command_output(Command::new("git").args(["ls-remote", "--tags", location]))
        .with_context(|| format!("failed to list remote tags for {location}"))?;
    let mut peeled = std::collections::BTreeMap::<String, String>::new();
    let mut direct = std::collections::BTreeMap::<String, String>::new();
    for line in output.lines() {
        let mut parts = line.split_whitespace();
        let Some(sha) = parts.next() else { continue };
        let Some(ref_name) = parts.next() else {
            continue;
        };
        let Some(tag) = ref_name.strip_prefix("refs/tags/") else {
            continue;
        };
        if let Some(tag) = tag.strip_suffix("^{}") {
            peeled.insert(tag.to_string(), sha.to_string());
        } else {
            direct.insert(tag.to_string(), sha.to_string());
        }
    }

    let mut versions = Vec::new();
    for (tag, sha) in direct {
        if let Some(version) = parse_swift_tag_version(&tag) {
            let revision = peeled.get(&tag).cloned().unwrap_or(sha);
            versions.push(RemoteVersion { version, revision });
        }
    }
    Ok(versions)
}

pub(crate) fn resolve_named_ref(location: &str, name: &str) -> Result<String> {
    let output = command_output(Command::new("git").args(["ls-remote", location, name]))
        .with_context(|| format!("failed to resolve {name} in {location}"))?;
    let line = output
        .lines()
        .next()
        .ok_or_else(|| anyhow!("{name} was not found in {location}"))?;
    let revision = line
        .split_whitespace()
        .next()
        .ok_or_else(|| anyhow!("ls-remote returned an invalid line"))?;
    Ok(revision.to_string())
}

pub(crate) fn parse_swift_tag_version(tag: &str) -> Option<Version> {
    let tag = tag.trim_start_matches('v');
    Version::parse(tag).ok()
}

File: Sources/swifterpm/Remote.swift
import Foundation

struct RemoteVersion: Codable, Sendable {
    let version: String
    let revision: String

    var semver: SemVer? {
        try? SemVer(version)
    }
}

private struct RemoteVersionsCache: Codable {
    let location: String
    let versions: [RemoteVersion]
}

func remoteVersions(location: String, cache: Cache) async throws -> [RemoteVersion] {
    if let cached = try readCachedRemoteVersions(cache: cache, location: location) {
        return cached
    }
    let lock = try cache.lock(namespace: "remote-versions", key: location)
    _ = lock
    if let cached = try readCachedRemoteVersions(cache: cache, location: location) {
        return cached
    }
    let versions = try await fetchRemoteVersions(location: location)
    try writeCachedRemoteVersions(cache: cache, location: location, versions: versions)
    return versions
}

private func fetchRemoteVersions(location: String) async throws -> [RemoteVersion] {
    let gitVersions = (try? gitRemoteVersions(location: location)) ?? []
    if !gitVersions.isEmpty {
        return gitVersions
    }
    guard let repo = try? GitHubRepo(location: location) else {
        return []
    }
    return try await githubRemoteVersions(repo: repo)
}

private func gitRemoteVersions(location: String) throws -> [RemoteVersion] {
    let output = try commandOutput("/usr/bin/git", ["ls-remote", "--tags", location])
    var peeled: [String: String] = [:]
    var direct: [String: String] = [:]
    for line in output.split(separator: "\n") {
        let parts = line.split(whereSeparator: \.isWhitespace)
        guard parts.count >= 2 else { continue }
        let sha = String(parts[0])
        let refName = String(parts[1])
        guard refName.hasPrefix("refs/tags/") else { continue }
        var tag = String(refName.dropFirst("refs/tags/".count))
        if tag.hasSuffix("^{}") {
            tag = String(tag.dropLast(3))
            peeled[tag] = sha
        } else {
            direct[tag] = sha
        }
    }

    var versions: [RemoteVersion] = []
    for (tag, sha) in direct {
        guard let version = parseSwiftTagVersion(tag) else { continue }
        versions.append(RemoteVersion(version: version.description, revision: peeled[tag] ?? sha))
    }
    return versions.sorted { (try? SemVer($0.version)) ?? SemVer(major: 0, minor: 0, patch: 0) < ((try? SemVer($1.version)) ?? SemVer(major: 0, minor: 0, patch: 0)) }
}

private func githubRemoteVersions(repo: GitHubRepo) async throws -> [RemoteVersion] {
    struct TagsResponse: Decodable {
        struct Commit: Decodable { let sha: String }
        let name: String
        let commit: Commit
    }

    var versions: [RemoteVersion] = []
    var page = 1
    while true {
        let url = URL(string: "https://api.github.com/repos/\(repo.owner)/\(repo.repo)/tags?per_page=100&page=\(page)")!
        var headers = ["User-Agent": "swifterpm/0.1"]
        if let token = githubToken() {
            headers["Authorization"] = "Bearer \(token)"
        }
        let tags = try JSONDecoder().decode([TagsResponse].self, from: await httpData(url: url, headers: headers))
        if tags.isEmpty {
            break
        }
        for tag in tags {
            if let version = parseSwiftTagVersion(tag.name) {
                versions.append(RemoteVersion(version: version.description, revision: tag.commit.sha))
            }
        }
        page += 1
    }
    return versions.sorted { (try? SemVer($0.version)) ?? SemVer(major: 0, minor: 0, patch: 0) < ((try? SemVer($1.version)) ?? SemVer(major: 0, minor: 0, patch: 0)) }
}

func resolveNamedRef(location: String, name: String) throws -> String {
    let output = try commandOutput("/usr/bin/git", ["ls-remote", location, name])
    guard let line = output.split(separator: "\n").first,
          let revision = line.split(whereSeparator: \.isWhitespace).first
    else {
        throw ToolError.message("\(name) was not found in \(location)")
    }
    return String(revision)
}

func parseSwiftTagVersion(_ tag: String) -> SemVer? {
    let value = tag.hasPrefix("v") ? String(tag.dropFirst()) : tag
    return try? SemVer(value)
}

private func readCachedRemoteVersions(cache: Cache, location: String) throws -> [RemoteVersion]? {
    let path = cache.remoteVersionsPath(location: location)
    guard FileManager.default.fileExists(atPath: path.path) else { return nil }
    let attrs = try FileManager.default.attributesOfItem(atPath: path.path)
    if let modified = attrs[.modificationDate] as? Date,
       Date().timeIntervalSince(modified) > 60 * 60
    {
        return nil
    }
    let cached = try JSONDecoder().decode(RemoteVersionsCache.self, from: Data(contentsOf: path))
    guard cached.location == location else { return nil }
    return cached.versions
}

private func writeCachedRemoteVersions(cache: Cache, location: String, versions: [RemoteVersion]) throws {
    let encoder = JSONEncoder()
    encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
    let data = try encoder.encode(RemoteVersionsCache(location: location, versions: versions)) + Data("\n".utf8)
    try atomicWrite(data, to: cache.remoteVersionsPath(location: location))
}
